Source/Core/DolphinQt2/Config/Mapping/IOWindow.cpp

```diff
@@ -0,0 +1,277 @@
+  connect(m_or_button, &QPushButton::clicked, this,
+          [this] { AppendSelectedOption(QStringLiteral(" | ")); });
+  connect(m_not_button, &QPushButton::clicked, this,
+          [this] { AppendSelectedOption(QStringLiteral("!")); });
+
+  connect(m_type == IOWindow::Type::Input ? m_detect_button : m_test_button, &QPushButton::clicked,
+          this, &IOWindow::OnDetectButtonPressed);
+
+  connect(m_button_box, &QDialogButtonBox::clicked, this, &IOWindow::OnDialogButtonPressed);
+  connect(m_devices_combo,
+          static_cast<void (QComboBox::*)(const QString&)>(&QComboBox::currentIndexChanged), this,
+          &IOWindow::OnDeviceChanged);
+  connect(m_range_spinbox, static_cast<void (QSpinBox::*)(int value)>(&QSpinBox::valueChanged),
+          this, &IOWindow::OnRangeChanged);
+  connect(m_range_slider, static_cast<void (QSlider::*)(int value)>(&QSlider::valueChanged), this,
+          &IOWindow::OnRangeChanged);
+}
+
+void IOWindow::AppendSelectedOption(const QString& prefix)
+{
+  if (m_option_list->currentItem() == nullptr)
+    return;
+
+  m_expression_text->insertPlainText(
+      MappingCommon::GetExpressionForControl(m_option_list->currentItem()->text(), &m_devq,
+                                             &m_controller->default_device)
+          .prepend(prefix));
+}
+
+void IOWindow::OnDeviceChanged(const QString& device)
+{
+  m_devq.FromString(device.toStdString());
+  UpdateOptionList();
+}
+
+void IOWindow::OnDialogButtonPressed(QAbstractButton* button)
+{
+  if (button == m_clear_button)
+  {
+    m_expression_text->clear();
+    return;
+  }
+  else
+  {
+    m_reference->expression = m_expression_text->toPlainText().toStdString();
+  }
+
+  if (button != m_apply_button)
+    accept();
+}
+
+void IOWindow::OnDetectButtonPressed()
+{
+  if (m_block)
+    return;
+
+  m_block = true;
+  std::thread([this] {
+    auto* btn = m_type == IOWindow::Type::Input ? m_detect_button : m_test_button;
+    const auto old_label = btn->text();
+
+    btn->setText(QStringLiteral("..."));
+
+    ciface::Core::Device::Control* const ctrl =
+        m_reference->Detect(5000, g_controller_interface.FindDevice(m_devq).get());
+
+    btn->setText(old_label);
+
+    if (ctrl)
+    {
+      const auto name = QString::fromStdString(ctrl->GetName());
+      const auto list = m_option_list->findItems(name, Qt::MatchFixedString);
+
+      if (list.size() > 0)
+        m_option_list->setCurrentItem(list[0]);
+    }
+    m_block = false;
+  }).detach();
+}
+
+void IOWindow::OnRangeChanged(int value)
+{
+  m_reference->range = static_cast<double>(value) / SLIDER_TICK_COUNT;
+  m_range_spinbox->setValue(m_reference->range * SLIDER_TICK_COUNT);
+  m_range_slider->setValue(m_reference->range * SLIDER_TICK_COUNT);
+}
+
+void IOWindow::UpdateOptionList()
+{
+  if (m_block)
+    return;
+
+  m_option_list->clear();
+
+  const auto device = g_controller_interface.FindDevice(m_devq);
+  if (m_reference->IsInput())
```

Comment:
If you wanted, you could deduplicate this with something like:

```c++
Control* control_iter = m_reference->IsInput() ? &device->Inputs().front() : &device->Outputs().front();
const Control* control_end = (m_reference->IsInput() ? &device->Inputs().back() : &device->Outputs().back()) + 1;
for (; control_iter != control_end; control_iter++)
{
  m_option_list->addItem(QString::fromStdString(control_iter->GetName()));
}
```

since inputs and outputs are both subclasses of `Device::Control`.